extract_feature_print.py

```diff
@@ -1,128 +1,110 @@
 
-# device=sys.argv[1]
-n_part = int(sys.argv[2])
-i_part = int(sys.argv[3])
-if len(sys.argv) == 6:
-    exp_dir = sys.argv[4]
-    version = sys.argv[5]
-else:
-    i_gpu = sys.argv[4]
-    exp_dir = sys.argv[5]
+def extract_feature_print(device, n_part, i_part, i_gpu, exp_dir, version):
     os.environ["CUDA_VISIBLE_DEVICES"] = str(i_gpu)
-    version = sys.argv[6]
-import torch
-import torch.nn.functional as F
-import soundfile as sf
-import numpy as np
-from fairseq import checkpoint_utils
-
-device = "cpu"
-if torch.cuda.is_available():
-    device = "cuda"
-elif torch.backends.mps.is_available():
-    device = "mps"
-
-f = open("%s/extract_f0_feature.log" % exp_dir, "a+")
-
-
-def printt(strr):
-    print(strr)
-    f.write("%s\n" % strr)
-    f.flush()
-
-
-printt(sys.argv)
-base_dir = Path(__file__).parent.parent.resolve()
-default_path = base_dir / "models"
-models_path = Path(os.environ.get("MODELS_PATH", str(default_path))).resolve()
-models_path = models_path / "voice-swap"
-model_path = str(models_path / "hubert_base.pt") 
-
-printt(exp_dir)
-wavPath = "%s/1_16k_wavs" % exp_dir
-outPath = (
-    "%s/3_feature256" % exp_dir if version == "v1" else "%s/3_feature768" % exp_dir
-)
-os.makedirs(outPath, exist_ok=True)
-
-
-# wave must be 16k, hop_size=320
-def readwave(wav_path, normalize=False):
-    wav, sr = sf.read(wav_path)
-    assert sr == 16000
-    feats = torch.from_numpy(wav).float()
-    if feats.dim() == 2:  # double channels
-        feats = feats.mean(-1)
-    assert feats.dim() == 1, feats.dim()
-    if normalize:
-        with torch.no_grad():
-            feats = F.layer_norm(feats, feats.shape)
-    feats = feats.view(1, -1)
-    return feats
-
-
-# HuBERT model
-printt("load model(s) from {}".format(model_path))
-# if hubert model is exist
-if os.access(model_path, os.F_OK) == False:
-    printt(
-        "Error: Extracting is shut down because %s does not exist, you may download it from https://huggingface.co/lj1995/VoiceConversionWebUI/tree/main"
-        % model_path
+    import torch
+    import torch.nn.functional as F
+    import soundfile as sf
+    import numpy as np
+    from fairseq import checkpoint_utils
+
+    device = "cpu"
+    if torch.cuda.is_available():
+        device = "cuda"
+    elif torch.backends.mps.is_available():
+        device = "mps"
+
+    f = open("%s/extract_f0_feature.log" % exp_dir, "a+")
+
+
+    base_dir = Path(__file__).parent.parent.resolve()
+    default_path = base_dir / "models"
+    models_path = Path(os.environ.get("MODELS_PATH", str(default_path))).resolve()
+    models_path = models_path / "voice-swap"
+    model_path = str(models_path / "hubert_base.pt") 
+
+    wavPath = "%s/1_16k_wavs" % exp_dir
+    outPath = (
+        "%s/3_feature256" % exp_dir if version == "v1" else "%s/3_feature768" % exp_dir
+    )
+    os.makedirs(outPath, exist_ok=True)
+
+    # wave must be 16k, hop_size=320
+    def readwave(wav_path, normalize=False):
+        wav, sr = sf.read(wav_path)

```

Comment:
what if it's not 16K?
do you resample before to 16k?

Reply:
This method works on all files in `f"{exp_dir}/1_16k_wavs/"`, which is created in the preprocessing phase:
```
        tmp_audio = librosa.resample(
            tmp_audio, orig_sr=self.sr, target_sr=16000
        )  # , res_type="soxr_vhq"
        wavfile.write(
            "%s/%s_%s.wav" % (self.wavs16k_dir, idx0, idx1),
            16000,
            tmp_audio.astype(np.float32),
        )
```